test/functional/eval/fnamemodify_spec.lua

```diff
@@ -36,4 +36,187 @@ describe('fnamemodify()', function()
+    end)
+
+    it(':h', function()
+      eq('src', fnamemodify(filename, ':h'))
+    end)
+
+    it(':t', function()
+      eq('version.c', fnamemodify(filename, ':t'))
+    end)
+
+    it(':r', function()
+      eq('src/version', fnamemodify(filename, ':r'))
+    end)
+
+    it(':t:r', function()
+      eq('version', fnamemodify(filename, ':t:r'))
+    end)
+
+    it(':e', function()
+      eq('c', fnamemodify(filename, ':e'))
+    end)
+
+    it(':s?version?main?', function()
+      eq('src/main.c', fnamemodify(filename, ':s?version?main?'))
+    end)
+  end)
+
+  describe('advanced examples from ":help filename-modifiers"', function()
+    local filename = "src/version.c.gz"
+
+    it(':e', function()
+      eq('gz', fnamemodify(filename, ':e'))
+    end)
+
+    it(':e:e', function()
+      eq('c.gz', fnamemodify(filename, ':e:e'))
+    end)
+
+    it(':e:e:e', function()
+      eq('c.gz', fnamemodify(filename, ':e:e:e'))
+    end)
+
+    it(':e:e:r', function()
+      eq('c', fnamemodify(filename, ':e:e:r'))
+    end)
+
+    it(':r', function()
+      eq('src/version.c', fnamemodify(filename, ':r'))
+    end)
+
+    it(':r:e', function()
+      eq('c', fnamemodify(filename, ':r:e'))
+    end)
+
+    it(':r:r', function()
+      eq('src/version', fnamemodify(filename, ':r:r'))
+    end)
+
+    it(':r:r:r', function()
+      eq('src/version', fnamemodify(filename, ':r:r:r'))
+    end)
+  end)
+
+  describe('modify with :h', function()
+    it('handles bare filenames', function()
+      eq('.', fnamemodify('hello.txt', ':h'))
+    end)
+
+    it('handles paths', function()
+      eq('path/to', fnamemodify('path/to/hello.txt', ':h'))
+    end)
+  end)
+
+  describe('modify with :t', function()
+    it('handles bare filenames', function()
+      eq('hello.txt', fnamemodify('hello.txt', ':t'))
+    end)
+
+    it('handles paths', function()
+      eq('hello.txt', fnamemodify('path/to/hello.txt', ':t'))
+    end)
+  end)
+
+  describe('modify with :r', function()
+    it('handles bare filenames', function()
+      eq('hello', fnamemodify('hello.txt', ':r'))
+    end)
+
+    it('handles paths', function()
+      eq('path/to/hello', fnamemodify('path/to/hello.txt', ':r'))
+    end)
+  end)
+
+  describe('modify with :e', function()
+    it('handles bare filenames', function()
+      eq('txt', fnamemodify('hello.txt', ':e'))
+    end)
+
+    it('handles paths', function()
+      eq('txt', fnamemodify('path/to/hello.txt', ':e'))
+    end)
+  end)
+
+  describe('modify with regex replacements', function()
+    it('handles a simple s///', function()
+      eq('content-there-here.txt', fnamemodify('content-here-here.txt', ':s/here/there/'))
+    end)
+
+    it('handles global', function()
+      eq('content-there-there.txt', fnamemodify('content-here-here.txt', ':gs/here/there/'))
+    end)
+  end)
+
+  it('modify with shell escape', function()
+    eq("'hello there! quote '\\'' newline\n'", fnamemodify("hello there! quote ' newline\n", ':S'))
```

Comment:
can use `[[` for a Lua "raw" string (no escaping):
```suggestion
  eq([['hello there! quote '\'' newline\n']], fnamemodify("hello there! quote ' newline\n", ':S'))
```

Reply:
Ah yeah, cheers

Reply:
Ah wait, I didn't do this because the `\n` at the end then isn't interpolated - which we still want (with the surrounding quotes being used to contain it).

Reply:
Right, my mistake.